psi4/src/psi4/libsapt_solver/exch-ind30.cc

```diff
@@ -314,5 +319,305 @@ double SAPT2p3::exch_ind30_3(double **sBS) {
+    CompleteInd30 -= C_DDOT(nb * ns, STS_bs->pointer()[0], 1, omega_bs->pointer()[0], 1);
+
+//All Omega-dependent contributions have been completed, now Xi-dependent (J/K) contributions
+
+    auto SCt_Na = std::make_shared<Matrix>("SCt_Na", nn, na);
+    auto SCt_Nb = std::make_shared<Matrix>("SCt_Nb", nn, nb);
+    C_DGEMM('N', 'T', nn, na, nr, 1.0, Ct_Kr->pointer()[0], nr, sAR->pointer()[0], nr, 0.0,
+            SCt_Na->pointer()[0], na);
+    C_DGEMM('N', 'T', nn, nb, ns, 1.0, Ct_Ks->pointer()[0], ns, sBS->pointer()[0], ns, 0.0,
+            SCt_Nb->pointer()[0], nb);
+
+    auto preXiAA = std::make_shared<Matrix>("preXiAA", nn, na);
+    auto preXiBA = std::make_shared<Matrix>("preXiBA", nn, nb);
+    auto preXiAB = std::make_shared<Matrix>("preXiAB", nn, na);
+    auto preXiBB = std::make_shared<Matrix>("preXiBB", nn, nb);
+    C_DGEMM('N', 'N', nn, na, na, 1.0, SCt_Na->pointer()[0], na, &Dp[0][0], na + nb, 0.0,
+            preXiAA->pointer()[0], na);
+    C_DGEMM('N', 'N', nn, nb, na, 1.0, SCt_Na->pointer()[0], na, &Dp[0][na], na + nb, 0.0,
+            preXiBA->pointer()[0], nb);
+    C_DGEMM('N', 'N', nn, na, nb, 1.0, SCt_Nb->pointer()[0], nb, &Dp[na][0], na + nb, 0.0,
+            preXiAB->pointer()[0], na);
+    C_DGEMM('N', 'N', nn, nb, nb, 1.0, SCt_Nb->pointer()[0], nb, &Dp[na][na], na + nb, 0.0,
+            preXiBB->pointer()[0], nb);
+
+    auto XiAA = std::make_shared<Matrix>("XiAA", nn, nn);
+    auto XiAB = std::make_shared<Matrix>("XiAB", nn, nn);
+    auto XiBA = std::make_shared<Matrix>("XiBA", nn, nn);
+    auto XiBB = std::make_shared<Matrix>("XiBB", nn, nn);
+    C_DGEMM('N', 'T', nn, nn, na, 1.0, CoccA_->pointer()[0], na, preXiAA->pointer()[0], na, 0.0,
+            XiAA->pointer()[0], nn);
+    C_DGEMM('N', 'T', nn, nn, na, 1.0, CoccA_->pointer()[0], na, preXiAB->pointer()[0], na, 0.0,
+            XiAB->pointer()[0], nn);
+    C_DGEMM('N', 'T', nn, nn, nb, 1.0, CoccB_->pointer()[0], nb, preXiBA->pointer()[0], nb, 0.0,
+            XiBA->pointer()[0], nn);
+    C_DGEMM('N', 'T', nn, nn, nb, 1.0, CoccB_->pointer()[0], nb, preXiBB->pointer()[0], nb, 0.0,
+            XiBB->pointer()[0], nn);
+
+    preXiBB->add(preXiBA);  //enough to calculate JK for the sum of the two
+
+    Cl.clear();
+    Cr.clear();
+    Cl.push_back(CoccB_);
+    Cr.push_back(preXiBB);
+    jk_->compute();
+
+    std::shared_ptr<Matrix> J_XiBB = J[0];
+    std::shared_ptr<Matrix> K_XiBB = K[0]->transpose();
+
+    CompleteInd30 += 4.0 * C_DDOT(nn * nn, XiAA->pointer()[0], 1, J_XiBB->pointer()[0], 1);
+    CompleteInd30 -= 2.0 * C_DDOT(nn * nn, XiAA->pointer()[0], 1, K_XiBB->pointer()[0], 1);
+    CompleteInd30 += 4.0 * C_DDOT(nn * nn, XiAB->pointer()[0], 1, J_XiBB->pointer()[0], 1);
+    CompleteInd30 -= 2.0 * C_DDOT(nn * nn, XiAB->pointer()[0], 1, K_XiBB->pointer()[0], 1);
+
+    e_exch_ind30_sinf_ = CompleteInd30 - e_ind30_;
+
+    Process::environment.globals["SAPT EXCH-IND30(S^INF) ENERGY"] = e_exch_ind30_sinf_;
```

Comment:
Q for Lori: Do we want to define the global? I know we're shifting to wavefunction variables.

Reply:
It's a lot simpler to switch over from P::e.globals to wfn module-by-module (or at least method-by-method). Looks like over libsapt_solver are still globals, so this new one is good as global.